WIP: add tmpdir to tests so pytest can run on clusters

nicolocin opened the pull request:
## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary
<!--- What does your code do? -->
To fix #366

## Checklist
<!--- Please, let us know if you need help-->
- [ ] All tests passing
- [x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project
(we are using `black`: you can `pip install pre-commit`,
run `pre-commit install` in the `pydra` directory
and `black` will be run automatically with each commit)

nicolocin commented:
Note to self - as of 6f8c3a7 these tests are still failing:

```python
FAILED pydra/engine/tests/test_submitter.py::test_wf_in_wf[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_submitter.py::test_wf2[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'

FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_inputspec_9[slurm-result_no_submitter] - AssertionError: assert 'file.txt' == 'file_copy.txt'
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_inputspec_9[slurm-result_submitter] - AssertionError: assert 'file.txt' == 'file_copy.txt'
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_inputspec_9a[slurm-result_no_submitter] - AssertionError: assert 'vast-storage_copy' == 'file_copy'
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_inputspec_9a[slurm-result_submitter] - AssertionError: assert 'vast-storage_copy' == 'file_copy'
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_inputspec_9b[slurm-result_no_submitter] - AssertionError: assert 'vast-storage' == 'file'
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_inputspec_9b[slurm-result_submitter] - AssertionError: assert 'vast-storage' == 'file'

FAILED pydra/engine/tests/test_shelltask_inputspec.py::test_shell_cmd_inputs_template_7 - AssertionError: assert ('executable '\n '/net/vast-storage.ib.cluster/scratch/Thu/nlo...
FAILED pydra/engine/tests/test_shelltask_inputspec.py::test_shell_cmd_inputs_template_7a - AssertionError: assert ('executable '\n '/net/vast-storage.ib.cluster/scratch/Thu/nl...
FAILED pydra/engine/tests/test_shelltask_inputspec.py::test_shell_cmd_inputs_template_7b - AssertionError: assert ('executable '\n '/net/vast-storage.ib.cluster/scratch/Thu/nl...
FAILED pydra/engine/tests/test_shelltask_inputspec.py::test_shell_cmd_inputs_template_8 - AssertionError: assert ('executable '\n '/net/vast-storage.ib.cluster/scratch/Thu/nlo...
FAILED pydra/engine/tests/test_shelltask_inputspec.py::test_shell_cmd_inputs_di - AssertionError: assert ('DenoiseImage -i '\n '/net/vast-storage.ib.cluster/scratch/Thu/nlo/tm...

FAILED pydra/engine/tests/test_workflow.py::test_wf_ndst_6[slurm] - AttributeError: 'State' object has no attribute 'final_combined_ind_mapping'
FAILED pydra/engine/tests/test_workflow.py::test_wf_3sernd_ndst_1[slurm] - AttributeError: 'State' object has no attribute 'inputs_ind'
FAILED pydra/engine/tests/test_workflow.py::test_wf_ndstLR_2a[slurm] - AttributeError: 'State' object has no attribute 'inputs_ind'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfinp_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfndupdate[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_st_1[slurm] - ValueError: Task wfnd raised an error
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_st_updatespl_1[slurm] - ValueError: Task wfnd raised an error
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_ndst_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_ndst_updatespl_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfst_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_st_2[slurm] - RuntimeError: Event loop is closed
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfst_2[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_ndst_3[slurm] - ValueError: Task wfnd raised an error
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfst_3[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wf_upstream_error3[slurm] - AssertionError: assert 'addvar1' in 'Event loop is closed'
FAILED pydra/engine/tests/test_workflow.py::test_wf_upstream_error5[slurm] - assert 'addvar1' in "'NoneType' object has no attribute 'errored'"
FAILED pydra/engine/tests/test_workflow.py::test_wf_upstream_error6[slurm] - assert 'addvar1' in "'NoneType' object has no attribute 'errored'"
FAILED pydra/engine/tests/test_workflow.py::test_wf_upstream_error9a[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
```

djarecka commented on pydra/engine/tests/test_boutiques.py:
```diff
@@ -26,12 +26,12 @@
     btask.inputs.infile = Infile
     btask.inputs.maskfile = maskfile
-    res = results_function(btask, plugin)
+    res = results_function(btask, plugin, tmpdir)
```
It will be better if you don't modify the `result_function`, just add `tmpdir` to the `btask` directly

djarecka replied:
you actually add `cache_dir` only to one of the function that is used as `results_function`

nicolocin replied:
Ah ok. I thought it's just more convenient and it's good practice to use tmpdir in general.  Though I can go back and modify the code again

djarecka replied:
As you could see `result_no_submitter` and `result_submitter` are really minimal functions that were design to run the task and return the result. I don't want to add anything extra (plus you only modified one of them).

And I don't really see any gain, since you have to change every test anyway.

nicolocin replied:
> you actually add `cache_dir` only to one of the function that is used as `results_function`

Not sure if I catch what you mean here?

djarecka replied:
`results_function` is a variable used in the `parametrize` field and it could be either `result_no_submitter` or `result_submitter`. You added `tmpdir`  to `results_function` that suggests that this is used in both  `result_no_submitter` and `result_submitter`, but this is not the case, you only modified `result_submitter`.

But I'd really keep these functions as the minimal functions that run the task either using `Submitter` explicitly or not. Everything else can be set only once, before calling `results_function`

nicolocin commented:
Solved path issue in cab4598, current errors:

```bash
============================================================================ short test summary info =============================================================================
FAILED pydra/engine/tests/test_shelltask.py::test_wf_shell_cmd_ndst_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_submitter.py::test_wf_in_wf[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_submitter.py::test_wf2[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_submitter.py::test_slurm_max_jobs - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wf_3sernd_ndst_1[slurm] - AttributeError: 'State' object has no attribute 'inputs_ind'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfinp_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfndupdate[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_st_1[slurm] - ValueError: Task wfnd raised an error
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_st_updatespl_1[slurm] - ValueError: Task wfnd raised an error
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_ndst_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_ndst_updatespl_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfst_1[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_st_2[slurm] - ValueError: Tasks ['wfnd'] raised an error
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfst_2[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_ndst_3[slurm] - ValueError: Task wfnd raised an error
FAILED pydra/engine/tests/test_workflow.py::test_wfasnd_wfst_3[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
FAILED pydra/engine/tests/test_workflow.py::test_wf_upstream_error5[slurm] - assert 'addvar1' in "'NoneType' object has no attribute 'errored'"
FAILED pydra/engine/tests/test_workflow.py::test_wf_upstream_error6[slurm] - assert 'addvar1' in "'NoneType' object has no attribute 'errored'"
FAILED pydra/engine/tests/test_workflow.py::test_wf_upstream_error9a[slurm] - AttributeError: 'NoneType' object has no attribute 'errored'
```

nicolocin commented:
Actually it seems like some tests in `test_workflow.py` only fail occasionally.   For example `test_wf_ndstLR_2a`

```bash
nlo[03:12][1.17][-95%]nlo@openmind7:nipype-pydra$ pytest -vvs --basetemp=/scratch/Thu/nlo/tmpdir pydra/engine/tests/test_workflow.py::test_wf_ndstLR_2a[slurm]
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.7.7, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /om2/user/nlo/miniconda/envs/pydra/bin/python3.7
cachedir: .pytest_cache
rootdir: /home/nlo/scripts/nipype-pydra
plugins: cov-2.10.1, env-0.6.2, forked-1.3.0, rerunfailures-9.1.1, xdist-1.34.0
collected 1 item

pydra/engine/tests/test_workflow.py::test_wf_ndstLR_2a[slurm] FAILED

==================================================================================== FAILURES ====================================================================================
____________________________________________________________________________ test_wf_ndstLR_2a[slurm] ____________________________________________________________________________

plugin = 'slurm', tmpdir = local('/net/vast-storage.ib.cluster/scratch/Thu/nlo/tmpdir/test_wf_ndstLR_2a_slurm_0')

    def test_wf_ndstLR_2a(plugin, tmpdir):
        """ Test workflow with 2 tasks, splitters on tasks levels
            The second task has splitter that has Left part (from previous state)
            and the Right part (it's onw outer splitter)
        """
        wf = Workflow(name="wf_ndst_3", input_spec=["x", "y", "z"])
        wf.add(add2(name="add2", x=wf.lzin.x).split("x"))
        wf.add(
            fun_addvar3(name="addvar", a=wf.add2.lzout.out, b=wf.lzin.y, c=wf.lzin.z).split(
                ["_add2", ["b", "c"]]
            )
        )
        wf.inputs.x = [1, 2, 3]
        wf.inputs.y = [10, 20]
        wf.inputs.z = [100, 200]
        wf.set_output([("out", wf.addvar.lzout.out)])
        wf.plugin = plugin
        wf.cache_dir = tmpdir

        with Submitter(plugin=plugin) as sub:
>           sub(wf)

pydra/engine/tests/test_workflow.py:1462:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydra/engine/submitter.py:57: in __call__
    self.loop.run_until_complete(self.submit_workflow(runnable, rerun=rerun))
/om2/user/nlo/miniconda/envs/pydra/lib/python3.7/asyncio/base_events.py:587: in run_until_complete
    return future.result()
pydra/engine/submitter.py:72: in submit_workflow
    await workflow._run(self, rerun=rerun)
pydra/engine/core.py:970: in _run
    result.output = self._collect_outputs()
pydra/engine/core.py:1052: in _collect_outputs
    val_out = val.get_value(self)
pydra/engine/specs.py:675: in get_value
    result = node.result(state_index=state_index)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pydra.engine.task.FunctionTask object at 0x2b4b5976ba10>, state_index = None, return_inputs = False

    def result(self, state_index=None, return_inputs=False):
        """
        Retrieve the outcomes of this particular task.

        Parameters
        ----------
        state_index : :obj: `int`
            index of the element for task with splitter and multiple states
        return_inputs : :obj: `bool`, :obj:`str`
            if True or "val" result is returned together with values of the input fields,
            if "ind" result is returned together with indices of the input fields

        Returns
        -------
        result :

        """
        # TODO: check if result is available in load_result and
        # return a future if not
        if self.errored:
            return Result(output=None, runtime=None, errored=True)
        if self.state:
            if state_index is None:
                # if state_index=None, collecting all results
                if self.state.combiner:
                    return self._combined_output(return_inputs=return_inputs)
                else:
                    results = []
>                   for ind in range(len(self.state.inputs_ind)):
E                   AttributeError: 'State' object has no attribute 'inputs_ind'

pydra/engine/core.py:639: AttributeError
============================================================================ short test summary info =============================================================================
FAILED pydra/engine/tests/test_workflow.py::test_wf_ndstLR_2a[slurm] - AttributeError: 'State' object has no attribute 'inputs_ind'
=============================================================================== 1 failed in 6.01s ================================================================================


nlo[03:16][2.07][-95%]nlo@openmind7:nipype-pydra$ pytest -vvs --basetemp=/scratch/Thu/nlo/tmpdir pydra/engine/tests/test_workflow.py::test_wf_ndstLR_2a[slurm]
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.7.7, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /om2/user/nlo/miniconda/envs/pydra/bin/python3.7
cachedir: .pytest_cache
rootdir: /home/nlo/scripts/nipype-pydra
plugins: cov-2.10.1, env-0.6.2, forked-1.3.0, rerunfailures-9.1.1, xdist-1.34.0
collected 1 item

pydra/engine/tests/test_workflow.py::test_wf_ndstLR_2a[slurm] PASSED

=============================================================================== 1 passed in 17.86s ===============================================================================
```

djarecka commented on pydra/engine/tests/test_shelltask.py:
```diff
@@ -279,7 +295,9 @@ def test_shell_cmd_inputspec_1(plugin, results_function, use_validator):
         input_spec=my_input_spec,
+        cache_dir=tmpdir,
     )
+    shelly.cache_dir = tmpdir
```
this probably is not needed if you added in line 298

djarecka commented on pydra/engine/tests/test_shelltask.py:
```diff
@@ -740,14 +774,15 @@ def test_shell_cmd_inputspec_5a_exception(plugin):
 
 @pytest.mark.parametrize("results_function", [result_no_submitter, result_submitter])
-def test_shell_cmd_inputspec_6(plugin, results_function):
+def test_shell_cmd_inputspec_6(plugin, results_function, tmpdir):
```
looks like you didn't use tmpdir in the test

djarecka commented on pydra/engine/tests/test_workflow.py:
```diff
@@ -48,7 +48,7 @@ def test_wf_name_conflict2():
 
 
-def test_wf_no_output(plugin):
+def test_wf_no_output(plugin, tmpdir):
```
you're not using this in the test
